### PR TITLE
Support filtering with almost any number of paths

### PR DESCRIPTION
**Pull request by JonasDov**
Closes https://github.com/iTwin/presentation/issues/722

Performance benchmark says that the test is about 1sec slower than with previous implementation. That is because new implementation puts 1k nodes under a single node (previously it was 500). When putting more and more nodes under a single node the filtering gets slower and slower. Ran the tests 10 times and got these results

|     | Previous implementation      |   New implementation with 500 paths    | New implementation with 1k paths      |
| ------------- | ------------- | ------------- | ------------- |
| Min | 16426.47 ms | 14686.84 ms  | 16929.78 ms |
| Max | 17432.14 ms | 16331.06 ms | 18527.86 ms |
| Average | 16950.287 ms | 15416.8 ms | 17756.909 ms |

### CODE REVIEW

**Review by grigasp — COMMENTED** [review]
Love how this simplifies the filtering logic, only got some small nitpicks

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 1]
```diff
@@ -360,6 +359,23 @@ export const ECSQL_COLUMN_NAME_FilterECInstanceId = "FilterECInstanceId";
 // eslint-disable-next-line @typescript-eslint/naming-convention
 export const ECSQL_COLUMN_NAME_FilterClassName = "FilterClassName";
 
+function getClassECInstanceIds(matchingFilters: Array<MatchedFilter<InstanceKey>>) {
+  const classNamePositions = new Map<string, number>();
+  const classNameECInstanceIds = new Array<{ className: string; ecInstanceIds: Id64String[] }>();
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts]: Not really, changed to use map

File: packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts [hunk 2]
```diff
@@ -377,21 +394,24 @@ export function applyECInstanceIdsFilter(
         ...(def.query.ctes ?? []),
         // note: generally we'd use `VALUES (1,1),(2,2)`, but that doesn't work in ECSQL (https://github.com/iTwin/itwinjs-backlog/issues/865),
         // so using UNION as a workaround
-        `FilteringInfo(ECInstanceId, IsFilterTarget, FilterTargetOptions, FilterClassName) AS (
-          ${matchingFilters
-            .map((mc) =>
-              mc.isFilterTarget
-                ? `VALUES (${mc.id.id}, 1, ${mc.filterTargetOptions ? `'${JSON.stringify(mc.filterTargetOptions)}'` : "CAST(NULL AS TEXT)"}, '${mc.id.className}')`
-                : `VALUES (${mc.id.id}, 0, CAST(NULL AS TEXT), '${mc.id.className}')`,
+        `FilteringInfo(ECInstanceId, FilterClassName) AS (
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts:390]: Removed

File: packages/hierarchies/src/test/HierarchyFiltering.test.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            type: "label-grouping",
+            label: "",
+          },
+          depth: 1,
+        },
+      },
+      { autoExpand: false },
+      undefined,
+    ];
+    it("returns correct result for different types of identifiers", () => {
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/test/HierarchyFiltering.test.ts]: Added another describe block

File: packages/hierarchies/src/test/imodel/FilteringHierarchyDefinition.test.ts [hunk 4]
```diff
@@ -940,7 +935,7 @@ describe("FilteringHierarchyDefinition", () => {
         const filteringFactory = createFilteringHierarchyDefinition({
           imodelAccess: { ...classHierarchyInspector, imodelKey: "test-imodel-key" },
           sourceFactory,
-          nodeIdentifierPaths: [[createTestGenericNodeKey({ id: "custom 2" })]],
+          nodeIdentifierPaths: [[createTestGenericNodeKey({ id: "custom 2" })], [createTestGenericNodeKey({ id: "custom 2" })]],
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/test/imodel/FilteringHierarchyDefinition.test.ts:937]: This isn't needed for the test, it is needed for test coverage. Specifically this line in `FilteringHierarchyDefinition.ts` (line 287):
 ```ts
 entry.filterTargetOptions = HierarchyFilteringPath.mergeOptions(entry.filterTargetOptions, options);
 ```

File: packages/hierarchies/src/test/imodel/FilteringHierarchyDefinition.test.ts [hunk 5]
```diff
@@ -1170,15 +1165,11 @@ describe("FilteringHierarchyDefinition", () => {
...
-                isFilterTarget: false,
-                childrenIdentifierPaths: [[{ className: filterPathClass2.fullName, id: "0x456" }]],
               },
               {
                 id: { className: filterPathClass1.fullName, id: "0x789" },
-                isFilterTarget: true,
-                childrenIdentifierPaths: [],
               },
-            ],
+            ] as Array<MatchedFilter<InstanceKey>>,
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/test/imodel/FilteringHierarchyDefinition.test.ts]: Changed what argument type `applyECInstanceIdsFilter` function accepts. Now its not needed.

File: packages/hierarchies/src/test/imodel/FilteringHierarchyDefinition.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/hierarchies/src/test/imodel/FilteringHierarchyDefinition.test.ts:937]: well, you're covering the line, but checking if it produced a correct result - the only thing it checks is that the line doesn't cause a crash. there should be a test that makes sure the line is not only covered, but also does what we expect.

[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchies/src/test/imodel/FilteringHierarchyDefinition.test.ts]: Added a test `applies correct filtering options to itself`

**Review by grigasp — APPROVED** [review]

**Pull request merged**